File: internal/scraper/scraper.go
package scraper

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type Scraper struct {
	urls     []string
	dataType string
}

func NewScraper(urls []string, dataType string) *Scraper {
	return &Scraper{
		urls:     urls,
		dataType: dataType,
	}
}

func (s *Scraper) Scrape(ctx context.Context) []string {
	resultChan := make(chan []string)
	errChan := make(chan error)

	for _, url := range s.urls {
		go s.fetchData(ctx, url, resultChan, errChan)
	}

	timeout := time.After(25 * time.Second)
	var results []string
	for i := 0; i < len(s.urls); i++ {
		select {
		case res := <-resultChan:
			results = append(results, res...)
		case err := <-errChan:
			fmt.Printf("Error scraping %s data: %s\n", s.dataType, err)
		case <-timeout:
			fmt.Println("Scraping timed out.")
			return results
		}
	}
	return results
}

func (s *Scraper) fetchData(ctx context.Context, url string, resultChan chan []string, errChan chan error) {
	fmt.Printf("Obteniendo %s de %s...\n", s.dataType, url)

	req, _ := http.NewRequest(http.MethodGet, url, nil)
	resp, err := http.DefaultClient.Do(req.WithContext(ctx))
	if err != nil {
		errChan <- err
		return
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		errChan <- err
		return
	}

	lines := strings.Split(string(body), "\n")
	var validLines []string
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if strings.Count(trimmed, ":") > 1 {
			trimmed = strings.Split(trimmed, ":")[0] + ":" + strings.Split(trimmed, ":")[1]
		}
		if trimmed != "" || (strings.Contains(url, "user-agents") && (!strings.Contains(trimmed, "Android") &&
			!strings.Contains(trimmed, "iPhone") &&
			!strings.Contains(trimmed, "iPad") &&
			!strings.Contains(trimmed, "compatible;"))) {

			validLines = append(validLines, trimmed)
		}
	}
	resultChan <- validLines
}

func ScrapeProxies() []string {
	urls := []string{
		// "https://raw.githubusercontent.com/proxifly/free-proxy-list/main/proxies/protocols/http/data.txt",
		// "https://raw.githubusercontent.com/proxifly/free-proxy-list/refs/heads/main/proxies/all/data.txt",
		"https://raw.githubusercontent.com/officialputuid/KangProxy/refs/heads/KangProxy/https/https.txt",
		"https://raw.githubusercontent.com/vakhov/fresh-proxy-list/refs/heads/master/https.txt",
		// "https://raw.githubusercontent.com/prxchk/proxy-list/main/http.txt",
		// "https://raw.githubusercontent.com/proxifly/free-proxy-list/main/proxies/protocols/http/data.txt",
		// "https://raw.githubusercontent.com/vakhov/fresh-proxy-list/master/http.txt",
		// "https://raw.githubusercontent.com/MuRongPIG/Proxy-Master/main/http.txt",
		// "https://raw.githubusercontent.com/ProxyScraper/ProxyScraper/main/http.txt",
	}
	scraper := NewScraper(urls, "proxies")
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	return scraper.Scrape(ctx)
}

func ScrapeUserAgents() []string {
	urls := []string{
		"https://gist.githubusercontent.com/pzb/b4b6f57144aea7827ae4/raw/cf847b76a142955b1410c8bcef3aabe221a63db1/user-agents.txt",
	}
	scraper := NewScraper(urls, "user-agents")
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	// Número máximo de intentos
	maxRetries := 3

	for attempt := 1; attempt <= maxRetries; attempt++ {
		result := scraper.Scrape(ctx)
		if len(result) > 0 {
			// Si la operación tiene éxito, retornar los datos
			return result
		}

		// Si la operación falla, esperar un momento antes de volver a intentar
		fmt.Printf("Intento %d fallido. Reintentando...\n", attempt)
		scraper = NewScraper(urls, "user-agents")
		time.Sleep(2 * time.Second)
	}

	// Si todos los intentos fallan, retornar una lista vacía o manejar el error de otra manera.
	return []string{}
}

File: internal/scraper/scraper.py
"""
Scraping de proxies y user agents
"""
import logging
import requests
from typing import List
import time

logger = logging.getLogger(__name__)

def scrape_proxies() -> List[str]:
    """Scraping de proxies desde fuentes públicas"""
    proxy_urls = [
        "https://raw.githubusercontent.com/officialputuid/KangProxy/refs/heads/KangProxy/https/https.txt",
        "https://raw.githubusercontent.com/vakhov/fresh-proxy-list/refs/heads/master/https.txt",
        # Comentadas las fuentes menos confiables
        # "https://raw.githubusercontent.com/proxifly/free-proxy-list/main/proxies/protocols/http/data.txt",
        # "https://raw.githubusercontent.com/proxifly/free-proxy-list/refs/heads/main/proxies/all/data.txt",
    ]
    
    all_proxies = []
    
    for url in proxy_urls:
        try:
            logger.info(f"Obteniendo proxies de {url}")
            response = requests.get(url, timeout=10)
            response.raise_for_status()
            
            lines = response.text.strip().split('\n')
            valid_proxies = []
            
            for line in lines:
                line = line.strip()
                if line and ':' in line:
                    # Limpiar formato de proxy si tiene múltiples ':'
                    parts = line.split(':')
                    if len(parts) >= 2:
                        proxy = f"{parts[0]}:{parts[1]}"
                        valid_proxies.append(proxy)
            
            all_proxies.extend(valid_proxies)
            logger.info(f"Obtenidos {len(valid_proxies)} proxies de {url}")
            
        except requests.RequestException as e:
            logger.warning(f"Error obteniendo proxies de {url}: {e}")
        except Exception as e:
            logger.error(f"Error inesperado con {url}: {e}")
    
    # Eliminar duplicados manteniendo orden
    unique_proxies = list(dict.fromkeys(all_proxies))
    logger.info(f"Total de proxies únicos obtenidos: {len(unique_proxies)}")
    
    return unique_proxies

def scrape_user_agents() -> List[str]:
    """Scraping de user agents"""
    user_agent_urls = [
        "https://gist.githubusercontent.com/pzb/b4b6f57144aea7827ae4/raw/cf847b76a142955b1410c8bcef3aabe221a63db1/user-agents.txt",
    ]
    
    all_user_agents = []
    max_retries = 3
    
    for url in user_agent_urls:
        for attempt in range(max_retries):
            try:
                logger.info(f"Obteniendo user agents de {url} (intento {attempt + 1})")
                response = requests.get(url, timeout=10)
                response.raise_for_status()
                
                lines = response.text.strip().split('\n')
                valid_user_agents = []
                
                for line in lines:
                    line = line.strip()
                    # Filtrar user agents móviles y problemáticos
                    if (line and 
                        not any(mobile in line for mobile in ['Android', 'iPhone', 'iPad', 'Mobile']) and
                        'Mozilla/' in line):
                        valid_user_agents.append(line)
                
                all_user_agents.extend(valid_user_agents)
                logger.info(f"Obtenidos {len(valid_user_agents)} user agents de {url}")
                break  # Éxito, salir del bucle de reintentos
                
            except requests.RequestException as e:
                logger.warning(f"Error obteniendo user agents de {url} (intento {attempt + 1}): {e}")
                if attempt < max_retries - 1:
                    time.sleep(2)  # Esperar antes del siguiente intento
            except Exception as e:
                logger.error(f"Error inesperado con {url}: {e}")
                break
    
    # Si no se obtuvieron user agents, usar valores por defecto
    if not all_user_agents:
        logger.warning("No se pudieron obtener user agents, usando valores por defecto")
        from ..config.config import DEFAULT_USER_AGENTS
        all_user_agents = DEFAULT_USER_AGENTS
    
    # Eliminar duplicados
    unique_user_agents = list(dict.fromkeys(all_user_agents))
    logger.info(f"Total de user agents únicos: {len(unique_user_agents)}")
    
    return unique_user_agents